### DISCUSSION TITLE
Why LinearGradientBrush animation throws error?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#11336] [Q&A] [answered]
The version I am using is 11.0.0 Preview 8.

Here is XAML:
```
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:AvaloniaApplication1">
	<Design.PreviewWith>
		<controls:TemplatedControl1 />
	</Design.PreviewWith>

	<Style Selector="controls|TemplatedControl1">
		<!-- Set Defaults -->
		<Setter Property="Template">
			<ControlTemplate>
				<Border
					Name="border"
					Background="{TemplateBinding Background}"
					BorderBrush="{TemplateBinding BorderBrush}"
					BorderThickness="{TemplateBinding BorderThickness}"
					CornerRadius="{TemplateBinding CornerRadius}"
					Padding="{TemplateBinding Padding}" Width="100" Height="100">
				</Border>
			</ControlTemplate>
		</Setter>
	</Style>
	<Style Selector="controls|TemplatedControl1:checked Border#border">
		<Style.Animations>
			<Animation Duration="0:0:1">
				<KeyFrame Cue="0%">
					<Setter Property="Background" Value="White"/>
				</KeyFrame>
				<KeyFrame Cue="100%">
					<Setter Property="Background">
						<LinearGradientBrush StartPoint="0%,0%" EndPoint="100%,100%">
							<GradientStop Color="Blue" Offset="0" />
							<GradientStop Color="Red" Offset="1" />
						</LinearGradientBrush>
					</Setter>
				</KeyFrame>
			</Animation>
		</Style.Animations>
	</Style>
	<Style Selector="controls|TemplatedControl1 Border#border">
		<Style.Animations>
			<Animation Duration="0:0:1">
				<KeyFrame Cue="0%">
					<Setter Property="Background">
						<LinearGradientBrush StartPoint="0%,0%" EndPoint="100%,100%">
							<GradientStop Color="Blue" Offset="0" />
							<GradientStop Color="Red" Offset="1" />
						</LinearGradientBrush>
					</Setter>
				</KeyFrame>
				<KeyFrame Cue="100%">
					<Setter Property="Background" Value="White"/>

				</KeyFrame>
			</Animation>
		</Style.Animations>
	</Style>
</Styles>
```

In Visual Studio output:

> 04:15:17.490 [Error]  Exception caught by FireAndForget
> System.AggregateException: One or more errors occurred. ---> System.NullReferenceException: Object reference not set to an instance of an object.
>    at AvaloniaVS.Services.PreviewerProcess.<SendAsync>d__38.MoveNext() in D:\a\1\s\AvaloniaVS.Shared\Services\PreviewerProcess.cs:line 389
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at AvaloniaVS.Services.PreviewerProcess.<OnMessageAsync>d__39.MoveNext() in D:\a\1\s\AvaloniaVS.Shared\Services\PreviewerProcess.cs:line 425
>    --- End of inner exception stack trace ---
> ---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object.
>    at AvaloniaVS.Services.PreviewerProcess.<SendAsync>d__38.MoveNext() in D:\a\1\s\AvaloniaVS.Shared\Services\PreviewerProcess.cs:line 389
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at AvaloniaVS.Services.PreviewerProcess.<OnMessageAsync>d__39.MoveNext() in D:\a\1\s\AvaloniaVS.Shared\Services\PreviewerProcess.cs:line 425<---

The StackTrace of the error:

> System.InvalidCastException
>   HResult=0x80004002
>   Message=Unable to cast object of type 'Avalonia.Media.Immutable.ImmutableSolidColorBrush' to type 'Avalonia.Media.IGradientBrush'.
>   Source=Avalonia.Base
>   StackTrace:
>    at Avalonia.Animation.AnimationInstance`1.UpdateNeutralValue()
>    at Avalonia.Animation.AnimationInstance`1.Subscribed()
>    at Avalonia.Reactive.SingleSubscriberObservableBase`1.Subscribe(IObserver`1 observer)
>    at Avalonia.PropertyStore.BindingEntryBase`2.Start(Boolean produceValue)
>    at Avalonia.PropertyStore.BindingEntryBase`2.Start()
>    at Avalonia.PropertyStore.ValueStore.AddBinding[T](StyledProperty`1 property, IObservable`1 source, BindingPriority priority)
>    at Avalonia.AvaloniaObject.Bind[T](StyledProperty`1 property, IObservable`1 source, BindingPriority priority)
>    at Avalonia.StyledProperty`1.RouteBind(AvaloniaObject target, IObservable`1 source, BindingPriority priority)
>    at Avalonia.AvaloniaObject.Bind(AvaloniaProperty property, IObservable`1 source, BindingPriority priority)
>    at Avalonia.Animation.Animators.GradientBrushAnimator.BindAnimation(Animatable control, IObservable`1 instance)
>    at Avalonia.Animation.Animators.Animator`1.Run(Animation animation, Animatable control, IClock clock, Action onComplete)
>    at Avalonia.Animation.DisposeAnimationInstanceSubject`1.System.IObserver<System.Boolean>.OnNext(Boolean matchVal)
>    at Avalonia.Reactive.LightweightObservableBase`1.PublishNext(T value)
>    at Avalonia.Reactive.LightweightSubject`1.OnNext(T value)
>    at Avalonia.Styling.StyleInstance.ApplyAnimations(AvaloniaObject control)
>    at Avalonia.Styling.StyleBase.Attach(StyledElement target, IStyleActivator activator, FrameType type)
>    at Avalonia.Styling.Style.TryAttach(StyledElement target, Object host, FrameType type)
>    at Avalonia.StyledElement.ApplyStyle(IStyle style, IStyleHost host, FrameType type)
>    at Avalonia.StyledElement.ApplyStyle(IStyle style, IStyleHost host, FrameType type)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyles(IStyleHost host)
>    at Avalonia.StyledElement.ApplyStyling()
>    at Avalonia.StyledElement.OnAttachedToLogicalTreeCore(LogicalTreeAttachmentEventArgs e)
>    at Avalonia.StyledElement.Avalonia.Controls.ISetLogicalParent.SetParent(ILogical parent)
>    at Avalonia.Controls.Primitives.TemplatedControl.ApplyTemplate()
>    at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
>    at Avalonia.Layout.Layoutable.Measure(Size availableSize)
>    at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding, Thickness borderThickness)
>    at Avalonia.Controls.Border.MeasureOverride(Size availableSize)
>    at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
>    at Avalonia.Layout.Layoutable.Measure(Size availableSize)
>    at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding, Thickness borderThickness)
>    at Avalonia.Controls.Presenters.ContentPresenter.MeasureOverride(Size availableSize)
>    at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
>    at Avalonia.Layout.Layoutable.Measure(Size availableSize)
>    at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding)
>    at Avalonia.Controls.Decorator.MeasureOverride(Size availableSize)
>    at Avalonia.Controls.Primitives.VisualLayerManager.MeasureOverride(Size availableSize)
>    at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
>    at Avalonia.Layout.Layoutable.Measure(Size availableSize)
>    at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
>    at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
>    at Avalonia.Layout.Layoutable.Measure(Size availableSize)
>    at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
>    at Avalonia.Controls.Window.MeasureOverride(Size availableSize)
>    at Avalonia.Controls.WindowBase.MeasureCore(Size availableSize)
>    at Avalonia.Layout.Layoutable.Measure(Size availableSize)
>    at Avalonia.Layout.LayoutManager.Measure(Layoutable control)
>    at Avalonia.Layout.LayoutManager.ExecuteInitialLayoutPass()
>    at Avalonia.Controls.Window.ShowCore(Window owner)
>    at Avalonia.Controls.Window.Show()
>    at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.ShowMainWindow()
>    at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
>    at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, ShutdownMode shutdownMode)
>    at AvaloniaApplication1.Program.Main(String[] args) in C:\Users\mywatermelon\Desktop\AvaloniaApplication1\Program.cs:line 12

![1](https://github.com/AvaloniaUI/Avalonia/assets/30720036/e1fc7dc1-f611-4034-97c5-da94c7cec236)

Source code:https://github.com/kyuranger/LinearGradientBrushAnimation

What's wrong with my code?

**Comment by maxkatz6** [comment]
Hi! What Avalonia version are you using?
I don't think gradients animations were supported in 0.10, expecialy mixed (when on value is solid brush and another is gradient).
In 11.0 both scenarios should be fine.

**Reply by kyuranger** [reply]
> Hi! What Avalonia version are you using? I don't think gradients animations were supported in 0.10, expecialy mixed (when on value is solid brush and another is gradient). In 11.0 both scenarios should be fine.

The version I am using is 11.0 Preview 8.

**Reply by maxkatz6** [reply]
Do you have full stack trace?

**Reply by timunie** [reply]
Works for me in the sandbox App. So probably a minimal sample would be useful as well as the stack trace.

**Reply by kyuranger** [reply]
> Do you have full stack trace?

I just updated my question, simplified my code as much as possible (now it only uses XAML), added the error content suggested by visual studio output and StackTrace. Please have a look.

**Reply by kyuranger** [reply]
> Works for me in the sandbox App. So probably a minimal sample would be useful as well as the stack trace.

I just updated my question, simplified my code as much as possible (now it only uses XAML), added the error content suggested by visual studio output and StackTrace. Please have a look.

**Reply by timunie** [reply]
@kyuranger there seems to be a reference to `FireAndForget` and I believe Avalonia doesn't use this method. Can it be that you're doing async stuff without calling `Dispatcher.UiThread.Post()`? Again, please share your minimal sample shown in the video with us. Thx.

**Reply by kyuranger** [reply]
> @kyuranger there seems to be a reference to `FireAndForget` and I believe Avalonia doesn't use this method. Can it be that you're doing async stuff without calling `Dispatcher.UiThread.Post()`? Again, please share your minimal sample shown in the video with us. Thx.

I just update my source code:https://github.com/kyuranger/LinearGradientBrushAnimation

This is a minimal sample with a brand-new project. I created this project by VS with 11.0 Preview5 template and then updated it to Preview8 version. As you see in my code, I never use any FireAndForget .Please have a look.

**Reply by timunie** [reply, marked as answer]
@kyuranger it's the solid color Background you set on the instance in MainWindow. If I replace it like shown below with another `GradientBrush`, it works for me: 

original:
https://github.com/kyuranger/LinearGradientBrushAnimation/blob/5352f8a589393daf7b493107648d0fd95957ee2b/AvaloniaApplication1/MainWindow.axaml#L10

modified:
```xml
<local:TemplatedControl1>
	<local:TemplatedControl1.Background>
		<LinearGradientBrush>
			<GradientStop Color="Blue" />
		</LinearGradientBrush>
	</local:TemplatedControl1.Background>
</local:TemplatedControl1>
```

**Reply by kyuranger** [reply]
> @kyuranger it's the solid color Background you set on the instance in MainWindow. If I replace it like shown below with another `GradientBrush`, it works for me:
> 
> original: https://github.com/kyuranger/LinearGradientBrushAnimation/blob/5352f8a589393daf7b493107648d0fd95957ee2b/AvaloniaApplication1/MainWindow.axaml#L10
> 
> modified:
> 
> ```
> <local:TemplatedControl1>
> 	<local:TemplatedControl1.Background>
> 		<LinearGradientBrush>
> 			<GradientStop Color="Blue" />
> 		</LinearGradientBrush>
> 	</local:TemplatedControl1.Background>
> </local:TemplatedControl1>
> ```

In LinearGradientBrush animation of WPF it can use `Background="Blue" ` directly without any error. Maybe there is something different between WPF and Avalonia. Now I have to solve it according to what you said.

**Reply by timunie** [reply]
You can also create a feature request for the future if you think it is a useful one. But for now you at least have a workaround.

**Reply by kyuranger** [reply]
> You can also create a feature request for the future if you think it is a useful one. But for now you at least have a workaround.

Thank you so much.
